README.md
# base58

[![PyPI Version][pypi-image]](https://pypi.python.org/pypi?name=base58&:action=display)
[![PyPI Downloads][pypi-downloads-image]](https://pypi.python.org/pypi?name=base58&:action=display)
[![Build Status][travis-image]](https://travis-ci.org/keis/base58)
[![Coverage Status][coveralls-image]](https://coveralls.io/r/keis/base58?branch=master)

Base58 and Base58Check implementation compatible with what is used by the
bitcoin network. Any other alternative alphabet (like the Ripple one) can be used.

Starting from version 2.0.0 **python2 is no longer supported** the 1.x series
will remain supported but no new features will be added.


## Command line usage

    $ printf "hello world" | base58
    StV1DL6CwTryKyV

    $ printf "hello world" | base58 -c
    3vQB7B6MrGQZaxCuFg4oh

    $ printf "3vQB7B6MrGQZaxCuFg4oh" | base58 -dc
    hello world

    $ printf "4vQB7B6MrGQZaxCuFg4oh" | base58 -dc
    Invalid checksum


## Module usage

    >>> import base58
    >>> base58.b58encode(b'hello world')
    b'StV1DL6CwTryKyV'
    >>> base58.b58decode(b'StV1DL6CwTryKyV')
    b'hello world'
    >>> base58.b58encode_check(b'hello world')
    b'3vQB7B6MrGQZaxCuFg4oh'
    >>> base58.b58decode_check(b'3vQB7B6MrGQZaxCuFg4oh')
    b'hello world'
    >>> base58.b58decode_check(b'4vQB7B6MrGQZaxCuFg4oh')
    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
      File "base58.py", line 89, in b58decode_check
        raise ValueError("Invalid checksum")
    ValueError: Invalid checksum
    # Use another alphabet. Here, using the built-in Ripple alphabet.
    >>> base58.b58encode(b'hello world', alphabet=base58.RIPPLE_ALPHABET)
    b'StVrDLaUATiyKyV'
    >>> base58.b58decode(b'StVrDLaUATiyKyV', alphabet=base58.RIPPLE_ALPHABET)
    b'hello world'


[pypi-image]: https://img.shields.io/pypi/v/base58.svg?style=flat
[pypi-downloads-image]: https://img.shields.io/pypi/dm/base58.svg?style=flat
[travis-image]: https://img.shields.io/travis/keis/base58.svg?style=flat
[coveralls-image]: https://img.shields.io/coveralls/keis/base58.svg?style=flat


CHANGELOG.md
## Changelog

### v2.1.1 / 2021-10-30
- [#67](https://github.com/keis/base58/pull/67) use github actions (@keis)
- [#66](https://github.com/keis/base58/pull/66) Escape illegal special characters in error message (@hukkin)

### v2.1.0 / 2021-01-09
- [#64](https://github.com/keis/base58/pull/64) Parametrise a few tests over alphabet (@keis)
- [#63](https://github.com/keis/base58/pull/63) Made it handle any base with passing alphabet. (#63) (@tanupoo)
- [#62](https://github.com/keis/base58/pull/62) Setup cfg (@keis)
- [#61](https://github.com/keis/base58/pull/61) Add support to ppc64le (@gururajrkatti)
- [#60](https://github.com/keis/base58/pull/60) Update README.md (@alloynetworks)
- [#59](https://github.com/keis/base58/pull/59) New alias for XRP alphabet (@alloynetworks)
- [#58](https://github.com/keis/base58/pull/58) Improve invalid character message (@keis)
- [#57](https://github.com/keis/base58/pull/57) Autofix for similar letters (@keis)
- [#56](https://github.com/keis/base58/pull/56) Add performance benchmarks using pytest-benchmark (@keis)
- [#53](https://github.com/keis/base58/pull/53) Performance optimizations (@kolomenkin)

### v2.0.1 / 2020-06-06
- [#55](https://github.com/keis/base58/pull/55) Include license file in source distribution (@synapticarbors)
- [#50](https://github.com/keis/base58/pull/50) Typecheck tests now that hamcrest has typing (@keis)
- [#49](https://github.com/keis/base58/pull/49) Replace custom exception assert with hamcrest utils (@keis)

### v2.0.0 / 2020-01-14
- [#47](https://github.com/keis/base58/pull/47) Distribute type data (PEP 561) (@hukkinj1)
- [#48](https://github.com/keis/base58/pull/48) Use setup.cfg for mypy conf (@hukkinj1)
- [#46](https://github.com/keis/base58/pull/46) Allow str input to b58encode_check (@keis)
- [#44](https://github.com/keis/base58/pull/44) Type annotate public API (@hukkinj1)
- [#43](https://github.com/keis/base58/pull/43) Drop python 2.7 and 3.4 support (#43) (@hukkinj1)
- [#42](https://github.com/keis/base58/pull/42) Test py3.7 and py3.8. Set python_requires in setup.py (@hukkinj1)
- [#41](https://github.com/keis/base58/pull/41) Fix encode functions' return types in README (@hukkinj1)
- [#40](https://github.com/keis/base58/pull/40) Enhance/add alphabet param (#40) (@dannywillems)
- [#39](https://github.com/keis/base58/pull/39) accept other bytes-like types (@fametrano)

### v1.0.3 / 2018-12-28
- [#37](https://github.com/keis/base58/pull/37) Update base58.py (@pdelteil)

### v1.0.2 / 2018-09-27
- [#34](https://github.com/keis/base58/pull/34) Add bumpversion config (@keis)

### v1.0.1 / 2018-09-25
- [#31](https://github.com/keis/base58/pull/31) Include tests in PyPI tarball (@dotlambda)

### v1.0.0 / 2018-04-20
- [#27](https://github.com/keis/base58/pull/27) Use bytes for both input and output (@keis)
- [#25](https://github.com/keis/base58/pull/25) Do not strip newline from input to cli (@keis)
- [#26](https://github.com/keis/base58/pull/26) Use tox and pytest to run tests (@keis)
- [#22](https://github.com/keis/base58/pull/22) Add the packaging metadata to build the base58 snap (@elopio)
- [#17](https://github.com/keis/base58/pull/17) improved error message (@fametrano)
- [#21](https://github.com/keis/base58/pull/21) clearer padding (@fametrano)

### v0.2.5 / 2017-04-04
- [#14](https://github.com/keis/base58/pull/14) Slight optimization, version increment (@gappleto97)
- [#12](https://github.com/keis/base58/pull/12) Add integer support (#12) (@gappleto97)
- [#10](https://github.com/keis/base58/pull/10) Add more test systems (@gappleto97)

### v0.2.4 / 2016-10-28
- [#8](https://github.com/keis/base58/pull/8) Package metadata (@keis)

### v0.2.3 / 2016-06-14
- [#6](https://github.com/keis/base58/pull/6) Improve error message when the type is not bytes (@keis, @zkanda)

### v0.2.2 / 2015-04-09
- [#3](https://github.com/keis/base58/pull/3) test round trips (@oconnor663)
- [#2](https://github.com/keis/base58/pull/2) fix encoding of empty string (@keis)


test.py
from contextlib import contextmanager
from itertools import product
from hamcrest import assert_that, equal_to, instance_of
from base58 import b58encode, b58decode, b58encode_check, b58decode_check, b58encode_int, b58decode_int, alphabet


class RaisesContext(object):
    pass


@contextmanager
def assert_raises(matcher=None, message=''):
    # Short hand for instance_of matcher
    if isinstance(matcher, (type,)):
        matcher = instance_of(matcher)

    context = RaisesContext()
    try:
        yield context
    except Exception as e:
        context.exception = e

    assert_that(context.exception, matcher, message)


def test_simple_encode():
    data = b58encode(b'hello world')
    assert_that(data, equal_to('StV1DL6CwTryKyV'))


def test_leadingz_encode():
    data = b58encode(b'\0\0hello world')
    assert_that(data, equal_to('11StV1DL6CwTryKyV'))


def test_encode_empty():
    data = b58encode(b'')
    assert_that(data, equal_to(''))


def test_simple_decode():
    data = b58decode('StV1DL6CwTryKyV')
    assert_that(data, equal_to(b'hello world'))


def test_simple_decode_bytes():
    data = b58decode(b'StV1DL6CwTryKyV')
    assert_that(data, equal_to(b'hello world'))


def test_leadingz_decode():
    data = b58decode('11StV1DL6CwTryKyV')
    assert_that(data, equal_to(b'\0\0hello world'))


def test_decode_empty():
    data = b58decode('1')
    assert_that(data, equal_to(b'\0'))


def test_check_identity():
    data = b'hello world'
    out = b58decode_check(b58encode_check(data))
    assert_that(out, equal_to(data))


def test_check_reverse_identity():
    data = 'StV1DL6CwTryKyV'
    out = b58encode_check(b58decode_check(data))
    assert_that(out, equal_to(data))


def test_check_leadingz_identity():
    data = b'\0\0hello world'
    out = b58decode_check(b58encode_check(data))
    assert_that(out, equal_to(data))


def test_check_leadingz_reverse_identity():
    data = '11StV1DL6CwTryKyV'
    out = b58encode_check(b58decode_check(data))
    assert_that(out, equal_to(data))


def test_check_failure():
    data = '3vQB7B6MrGQZaxCuFg4oH'
    with assert_raises(ValueError):
        b58decode_check(data)


def test_round_trips():
    possible_bytes = [b'\x00', b'\x01', b'\x10', b'\xff']
    for length in range(0, 5):
        for bytes_to_test in product(possible_bytes, repeat=length):
            bytes_in = b''.join(bytes_to_test)
            bytes_out = b58decode(b58encode(bytes_in))
            assert_that(bytes_in, equal_to(bytes_out))


def test_input_should_be_bytes():
     data = u'3vQB7B6MrGQZaxCuFg4oH'
     with assert_raises(TypeError):
         b58encode(data)


def test_simple_integers():
    for idx, char in enumerate(alphabet):
        assert_that(b58decode_int(char), equal_to(idx))
        assert_that(b58encode_int(idx), equal_to(char))

def test_large_integer():
    number = 0x111d38e5fc9071ffcd20b4a763cc9ae4f252bb4e48fd66a835e252ada93ff480d6dd43dc62a641155a5
    assert_that(b58decode_int(alphabet), equal_to(number))
    assert_that(b58encode_int(number), equal_to(alphabet[1:]))


test_base45.py
from hamcrest import assert_that, equal_to, calling, raises
from base58 import (b58encode, b58decode, b58encode_check, b58decode_check)

BASE45_ALPHABET = b"0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ $%*+-./:"


def test_simple_encode():
    data = b58encode(b'hello world', alphabet=BASE45_ALPHABET)
    assert_that(data, equal_to(b'K3*J+EGLBVAYYB36'))


def test_leadingz_encode():
    data = b58encode(b'\0\0hello world', alphabet=BASE45_ALPHABET)
    assert_that(data, equal_to(b'00K3*J+EGLBVAYYB36'))


def test_encode_empty():
    data = b58encode(b'', alphabet=BASE45_ALPHABET)
    assert_that(data, equal_to(b''))


def test_simple_decode():
    data = b58decode('K3*J+EGLBVAYYB36', alphabet=BASE45_ALPHABET)
    assert_that(data, equal_to(b'hello world'))


def test_simple_decode_bytes():
    data = b58decode(b'K3*J+EGLBVAYYB36', alphabet=BASE45_ALPHABET)
    assert_that(data, equal_to(b'hello world'))


def test_autofix_decode_bytes():
    data = b58decode(
        b'K3*J+EGLBVAYYB36', alphabet=BASE45_ALPHABET, autofix=True)
    assert_that(data, equal_to(b'hello world'))


def test_leadingz_decode():
    data = b58decode('00K3*J+EGLBVAYYB36', alphabet=BASE45_ALPHABET)
    assert_that(data, equal_to(b'\0\0hello world'))


def test_leadingz_decode_bytes():
    data = b58decode(b'00K3*J+EGLBVAYYB36', alphabet=BASE45_ALPHABET)
    assert_that(data, equal_to(b'\0\0hello world'))


def test_empty_decode():
    data = b58decode('1', alphabet=BASE45_ALPHABET)
    assert_that(data, equal_to(b'\x01'))


def test_empty_decode_bytes():
    data = b58decode(b'1', alphabet=BASE45_ALPHABET)
    assert_that(data, equal_to(b'\x01'))


def test_check_str():
    data = 'hello world'
    out = b58encode_check(data, alphabet=BASE45_ALPHABET)
    assert_that(out, equal_to(b'AHN49RN6G8B%AWUALA8K2D'))
    back = b58decode_check(out, alphabet=BASE45_ALPHABET)
    assert_that(back, equal_to(b'hello world'))


def test_autofix_check_str():
    data = 'AHN49RN6G8B%AWUALA8K2D'
    back = b58decode_check(data, alphabet=BASE45_ALPHABET, autofix=True)
    assert_that(back, equal_to(b'hello world'))


def test_autofix_not_applicable_check_str():
    charset = BASE45_ALPHABET.replace(b'x', b'l')
    msg = b'hello world'
    enc = b58encode_check(msg, alphabet=BASE45_ALPHABET)
    modified = enc.replace(b'x', b'l').replace(b'o', b'0')
    back = b58decode_check(modified, alphabet=charset, autofix=True)
    assert_that(back, equal_to(msg))


def test_check_failure():
    data = '3vQB7B6MrGQZaxCuFg4oH'
    assert_that(calling(b58decode_check).with_args(data), raises(ValueError))


def test_invalid_input():
    data = 'xyz0'   # 0 is not part of the bitcoin base58 alphabet
    assert_that(
        calling(b58decode).with_args(data),
        raises(ValueError, "Invalid character '0'"))


test_base58.py
import pytest
from itertools import product
from random import getrandbits
from hamcrest import assert_that, equal_to, calling, raises
from base58 import (
    b58encode, b58decode, b58encode_check, b58decode_check, b58encode_int,
    b58decode_int,
    BITCOIN_ALPHABET,
    XRP_ALPHABET)


BASE45_ALPHABET = b"0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ $%*+-./:"


@pytest.fixture(params=[BITCOIN_ALPHABET, XRP_ALPHABET, BASE45_ALPHABET])
def alphabet(request) -> str:
    return request.param


def test_simple_encode():
    data = b58encode(b'hello world')
    assert_that(data, equal_to(b'StV1DL6CwTryKyV'))


def test_leadingz_encode():
    data = b58encode(b'\0\0hello world')
    assert_that(data, equal_to(b'11StV1DL6CwTryKyV'))


def test_encode_empty():
    data = b58encode(b'')
    assert_that(data, equal_to(b''))


def test_simple_decode():
    data = b58decode('StV1DL6CwTryKyV')
    assert_that(data, equal_to(b'hello world'))


def test_simple_decode_bytes():
    data = b58decode(b'StV1DL6CwTryKyV')
    assert_that(data, equal_to(b'hello world'))


def test_autofix_decode_bytes():
    data = b58decode(b'StVlDL6CwTryKyV', autofix=True)
    assert_that(data, equal_to(b'hello world'))


def test_leadingz_decode():
    data = b58decode('11StV1DL6CwTryKyV')
    assert_that(data, equal_to(b'\0\0hello world'))


def test_leadingz_decode_bytes():
    data = b58decode(b'11StV1DL6CwTryKyV')
    assert_that(data, equal_to(b'\0\0hello world'))


def test_empty_decode():
    data = b58decode('1')
    assert_that(data, equal_to(b'\0'))


def test_empty_decode_bytes():
    data = b58decode(b'1')
    assert_that(data, equal_to(b'\0'))


def test_check_str():
    data = 'hello world'
    out = b58encode_check(data)
    assert_that(out, equal_to(b'3vQB7B6MrGQZaxCuFg4oh'))
    back = b58decode_check(out)
    assert_that(back, equal_to(b'hello world'))


def test_autofix_check_str():
    data = '3vQB7B6MrGQZaxCuFg4Oh'
    back = b58decode_check(data, autofix=True)
    assert_that(back, equal_to(b'hello world'))


def test_autofix_not_applicable_check_str():
    charset = BITCOIN_ALPHABET.replace(b'x', b'l')
    msg = b'hello world'
    enc = b58encode_check(msg).replace(b'x', b'l').replace(b'o', b'0')
    back = b58decode_check(enc, alphabet=charset, autofix=True)
    assert_that(back, equal_to(msg))


def test_check_failure():
    data = '3vQB7B6MrGQZaxCuFg4oH'
    assert_that(calling(b58decode_check).with_args(data), raises(ValueError))


def test_check_identity(alphabet):
    data = b'hello world'
    out = b58decode_check(
        b58encode_check(data, alphabet=alphabet),
        alphabet=alphabet
    )
    assert_that(out, equal_to(data))


def test_round_trips(alphabet):
    possible_bytes = [b'\x00', b'\x01', b'\x10', b'\xff']
    for length in range(0, 5):
        for bytes_to_test in product(possible_bytes, repeat=length):
            bytes_in = b''.join(bytes_to_test)
            bytes_out = b58decode(
                b58encode(bytes_in, alphabet=alphabet),
                alphabet=alphabet)
            assert_that(bytes_in, equal_to(bytes_out))


def test_simple_integers(alphabet):
    for idx, char in enumerate(alphabet):
        charbytes = bytes([char])
        assert_that(b58decode_int(charbytes, alphabet=alphabet), equal_to(idx))
        assert_that(b58encode_int(idx, alphabet=alphabet), equal_to(charbytes))


def test_large_integer():
    number = 0x111d38e5fc9071ffcd20b4a763cc9ae4f252bb4e48fd66a835e252ada93ff480d6dd43dc62a641155a5  # noqa
    assert_that(b58decode_int(BITCOIN_ALPHABET), equal_to(number))
    assert_that(b58encode_int(number), equal_to(BITCOIN_ALPHABET[1:]))


def test_invalid_input():
    data = 'xyz\b'   # backspace is not part of the bitcoin base58 alphabet
    assert_that(
        calling(b58decode).with_args(data),
        raises(ValueError, "Invalid character '\\\\x08'"))


@pytest.mark.parametrize('length', [8, 32, 256, 1024])
def test_encode_random(benchmark, length) -> None:
    data = getrandbits(length * 8).to_bytes(length, byteorder='big')
    encoded = benchmark(lambda: b58encode(data))
    assert_that(b58decode(encoded), equal_to(data))


@pytest.mark.parametrize('length', [8, 32, 256, 1024])
def test_decode_random(benchmark, length) -> None:
    origdata = getrandbits(length * 8).to_bytes(length, byteorder='big')
    encoded = b58encode(origdata)
    data = benchmark(lambda: b58decode(encoded))
    assert_that(data, equal_to(origdata))


base58/__init__.py
'''Base58 encoding

Implementations of Base58 and Base58Check encodings that are compatible
with the bitcoin network.
'''

# This module is based upon base58 snippets found scattered over many bitcoin
# tools written in python. From what I gather the original source is from a
# forum post by Gavin Andresen, so direct your praise to him.
# This module adds shiny packaging and support for python3.

from functools import lru_cache
from hashlib import sha256
from typing import Mapping, Union

__version__ = '2.1.0'

# 58 character alphabet used
BITCOIN_ALPHABET = \
    b'123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz'
RIPPLE_ALPHABET = b'rpshnaf39wBUDNEGHJKLM4PQRST7VWXYZ2bcdeCg65jkm8oFqi1tuvAxyz'
XRP_ALPHABET = RIPPLE_ALPHABET

# Retro compatibility
alphabet = BITCOIN_ALPHABET


def scrub_input(v: Union[str, bytes]) -> bytes:
    if isinstance(v, str):
        v = v.encode('ascii')

    return v


def b58encode_int(
    i: int, default_one: bool = True, alphabet: bytes = BITCOIN_ALPHABET
) -> bytes:
    """
    Encode an integer using Base58
    """
    if not i and default_one:
        return alphabet[0:1]
    string = b""
    base = len(alphabet)
    while i:
        i, idx = divmod(i, base)
        string = alphabet[idx:idx+1] + string
    return string


def b58encode(
    v: Union[str, bytes], alphabet: bytes = BITCOIN_ALPHABET
) -> bytes:
    """
    Encode a string using Base58
    """
    v = scrub_input(v)

    origlen = len(v)
    v = v.lstrip(b'\0')
    newlen = len(v)

    acc = int.from_bytes(v, byteorder='big')  # first byte is most significant

    result = b58encode_int(acc, default_one=False, alphabet=alphabet)
    return alphabet[0:1] * (origlen - newlen) + result


@lru_cache()
def _get_base58_decode_map(alphabet: bytes,
                           autofix: bool) -> Mapping[int, int]:
    invmap = {char: index for index, char in enumerate(alphabet)}

    if autofix:
        groups = [b'0Oo', b'Il1']
        for group in groups:
            pivots = [c for c in group if c in invmap]
            if len(pivots) == 1:
                for alternative in group:
                    invmap[alternative] = invmap[pivots[0]]

    return invmap


def b58decode_int(
    v: Union[str, bytes], alphabet: bytes = BITCOIN_ALPHABET, *,
    autofix: bool = False
) -> int:
    """
    Decode a Base58 encoded string as an integer
    """
    if b' ' not in alphabet:
        v = v.rstrip()
    v = scrub_input(v)

    map = _get_base58_decode_map(alphabet, autofix=autofix)

    decimal = 0
    base = len(alphabet)
    try:
        for char in v:
            decimal = decimal * base + map[char]
    except KeyError as e:
        raise ValueError(
            "Invalid character {!r}".format(chr(e.args[0]))
        ) from None
    return decimal


def b58decode(
    v: Union[str, bytes], alphabet: bytes = BITCOIN_ALPHABET, *,
    autofix: bool = False
) -> bytes:
    """
    Decode a Base58 encoded string
    """
    v = v.rstrip()
    v = scrub_input(v)

    origlen = len(v)
    v = v.lstrip(alphabet[0:1])
    newlen = len(v)

    acc = b58decode_int(v, alphabet=alphabet, autofix=autofix)

    result = []
    while acc > 0:
        acc, mod = divmod(acc, 256)
        result.append(mod)

    return b'\0' * (origlen - newlen) + bytes(reversed(result))


def b58encode_check(
    v: Union[str, bytes], alphabet: bytes = BITCOIN_ALPHABET
) -> bytes:
    """
    Encode a string using Base58 with a 4 character checksum
    """
    v = scrub_input(v)

    digest = sha256(sha256(v).digest()).digest()
    return b58encode(v + digest[:4], alphabet=alphabet)


def b58decode_check(
    v: Union[str, bytes], alphabet: bytes = BITCOIN_ALPHABET, *,
    autofix: bool = False
) -> bytes:
    '''Decode and verify the checksum of a Base58 encoded string'''

    result = b58decode(v, alphabet=alphabet, autofix=autofix)
    result, check = result[:-4], result[-4:]
    digest = sha256(sha256(result).digest()).digest()

    if check != digest[:4]:
        raise ValueError("Invalid checksum")

    return result


base58/__main__.py
import argparse
import sys
from typing import Callable, Dict, Tuple

from base58 import b58decode, b58decode_check, b58encode, b58encode_check

_fmap: Dict[Tuple[bool, bool], Callable[[bytes], bytes]] = {
    (False, False): b58encode,
    (False, True): b58encode_check,
    (True, False): b58decode,
    (True, True): b58decode_check
}


def main() -> None:
    '''Base58 encode or decode FILE, or standard input, to standard output.'''

    stdout = sys.stdout.buffer

    parser = argparse.ArgumentParser(description=main.__doc__)
    parser.add_argument(
        'file',
        metavar='FILE',
        nargs='?',
        type=argparse.FileType('r'),
        default='-')
    parser.add_argument(
        '-d', '--decode',
        action='store_true',
        help='decode data')
    parser.add_argument(
        '-c', '--check',
        action='store_true',
        help='append a checksum before encoding')

    args = parser.parse_args()
    fun = _fmap[(args.decode, args.check)]

    data = args.file.buffer.read()

    try:
        result = fun(data)
    except Exception as e:
        sys.exit(e)

    stdout.write(result)


if __name__ == '__main__':
    main()


base58.py
'''Base58 encoding

Implementations of Base58 and Base58Check endcodings that are compatible
with the bitcoin network.
'''

# This module is based upon base58 snippets found scattered over many bitcoin
# tools written in python. From what I gather the original source is from a
# forum post by Gavin Andresen, so direct your praise to him.
# This module adds shiny packaging and support for python3.

from hashlib import sha256
from typing import Union

__version__ = '1.0.3'

# 58 character alphabet used
BITCOIN_ALPHABET = \
    b'123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz'
RIPPLE_ALPHABET = b'rpshnaf39wBUDNEGHJKLM4PQRST7VWXYZ2bcdeCg65jkm8oFqi1tuvAxyz'

# Retro compatibility
alphabet = BITCOIN_ALPHABET


def scrub_input(v: Union[str, bytes]) -> bytes:
    if isinstance(v, str):
        v = v.encode('ascii')

    return v


def b58encode_int(
    i: int, default_one: bool = True, alphabet: bytes = BITCOIN_ALPHABET
) -> bytes:
    """
    Encode an integer using Base58
    """
    if not i and default_one:
        return alphabet[0:1]
    string = b""
    while i:
        i, idx = divmod(i, 58)
        string = alphabet[idx:idx+1] + string
    return string


def b58encode(
    v: Union[str, bytes], alphabet: bytes = BITCOIN_ALPHABET
) -> bytes:
    """
    Encode a string using Base58
    """
    v = scrub_input(v)

    nPad = len(v)
    v = v.lstrip(b'\0')
    nPad -= len(v)

    p, acc = 1, 0
    for c in reversed(v):
        acc += p * c
        p = p << 8
    result = b58encode_int(acc, default_one=False, alphabet=alphabet)
    return alphabet[0:1] * nPad + result


def b58decode_int(
    v: Union[str, bytes], alphabet: bytes = BITCOIN_ALPHABET
) -> int:
    """
    Decode a Base58 encoded string as an integer
    """
    v = v.rstrip()
    v = scrub_input(v)

    decimal = 0
    for char in v:
        decimal = decimal * 58 + alphabet.index(char)
    return decimal


def b58decode(
    v: Union[str, bytes], alphabet: bytes = BITCOIN_ALPHABET
) -> bytes:
    """
    Decode a Base58 encoded string
    """
    v = v.rstrip()
    v = scrub_input(v)

    origlen = len(v)
    v = v.lstrip(alphabet[0:1])
    newlen = len(v)

    acc = b58decode_int(v, alphabet=alphabet)

    result = []
    while acc > 0:
        acc, mod = divmod(acc, 256)
        result.append(mod)

    return b'\0' * (origlen - newlen) + bytes(reversed(result))


def b58encode_check(
    v: Union[str, bytes], alphabet: bytes = BITCOIN_ALPHABET
) -> bytes:
    """
    Encode a string using Base58 with a 4 character checksum
    """
    v = scrub_input(v)

    digest = sha256(sha256(v).digest()).digest()
    return b58encode(v + digest[:4], alphabet=alphabet)


def b58decode_check(
    v: Union[str, bytes], alphabet: bytes = BITCOIN_ALPHABET
) -> bytes:
    '''Decode and verify the checksum of a Base58 encoded string'''

    result = b58decode(v, alphabet=alphabet)
    result, check = result[:-4], result[-4:]
    digest = sha256(sha256(result).digest()).digest()

    if check != digest[:4]:
        raise ValueError("Invalid checksum")

    return result


def main():
    '''Base58 encode or decode FILE, or standard input, to standard output.'''

    import sys
    import argparse

    stdout = sys.stdout.buffer

    parser = argparse.ArgumentParser(description=main.__doc__)
    parser.add_argument(
        'file',
        metavar='FILE',
        nargs='?',
        type=argparse.FileType('r'),
        default='-')
    parser.add_argument(
        '-d', '--decode',
        action='store_true',
        help='decode data')
    parser.add_argument(
        '-c', '--check',
        action='store_true',
        help='append a checksum before encoding')

    args = parser.parse_args()
    fun = {
        (False, False): b58encode,
        (False, True): b58encode_check,
        (True, False): b58decode,
        (True, True): b58decode_check
    }[(args.decode, args.check)]

    data = args.file.buffer.read()

    try:
        result = fun(data)
    except Exception as e:
        sys.exit(e)

    stdout.write(result)


if __name__ == '__main__':
    main()
